packages/labs/src/inputList.tsx

```diff
@@ -0,0 +1,293 @@
+     * See `items` for the full unfiltered list.
+     */
+    filteredItems: T[];
+
+    /**
+     * Selection handler that should be invoked when a new item has been chosen,
+     * perhaps because the user clicked it.
+     */
+    handleItemSelect: (item: T, event: React.SyntheticEvent<HTMLElement>) => void;
+
+    /**
+     * Keyboard handler for up/down arrow keys to shift the active item.
+     * Attach this handler to any element that should support this interaction.
+     */
+    handleKeyDown: React.KeyboardEventHandler<HTMLElement>;
+
+    /**
+     * Keyboard handler for enter key to select the active item.
+     * Attach this handler to any element that should support this interaction.
+     */
+    handleKeyUp: React.KeyboardEventHandler<HTMLElement>;
+
+    /**
+     * Array of all (unfiltered) items in the list.
+     * See `filteredItems` for a filtered array based on `query`.
+     */
+    items: T[];
+
+    /**
+     * A ref handler that should be applied to the HTML element that contains the rendererd items.
+     * This is required for the `InputList` to scroll the active item into view automatically.
+     */
+    itemsParentRef: (ref: HTMLElement) => void;
+
+    /**
+     * Controlled text value of the filter input. Attach an `onChange` handler to the relevant
+     * element to control this prop from your application's state.
+     */
+    query: string;
+}
+
+export interface IInputListState<T> {
+    filteredItems?: T[];
+}
+
+export class InputList<T> extends React.Component<IInputListProps<T>, IInputListState<T>> {
+    public static displayName = "Blueprint.InputList";
+
+    public static ofType<T>() {
+        return InputList as new (props: IInputListProps<T>) => InputList<T>;
+    }
+
+    private itemsParentRef: HTMLElement;
+    private refHandlers = {
+        itemsParent: (ref: HTMLElement) => this.itemsParentRef = ref,
+    };
+
+    /**
+     * flag indicating that we should check whether selected item is in viewport after rendering,
+     * typically because of keyboard change.
+     */
+    private shouldCheckActiveItemInViewport: boolean;
+
+    public render() {
+        const { renderer, ...props } = this.props;
+        const { filteredItems } = this.state;
+        return renderer({
+            ...props,
+            filteredItems,
+            handleItemSelect: this.handleItemSelect,
+            handleKeyDown: this.handleKeyDown,
+            handleKeyUp: this.handleKeyUp,
+            itemsParentRef: this.refHandlers.itemsParent,
+        });
+    }
+
+    public componentWillMount() {
+        this.setState({ filteredItems: getFilteredItems(this.props) });
+    }
+
+    public componentWillReceiveProps(nextProps: IInputListProps<T>) {
+        if (nextProps.items !== this.props.items
+            || nextProps.itemListPredicate !== this.props.itemListPredicate
+            || nextProps.itemPredicate !== this.props.itemPredicate
+            || nextProps.query !== this.props.query
```

Comment:
Let's use `Utils.shallowCompareKeys` eventually, or now if you already have it on this branch.